# Supermarket Loyalty

International Essentials is an international supermarket chain.

Shoppers at their supermarkets can sign up for a loyalty program that provides rewards each year to customers based on their spending. The more you spend the bigger the rewards. 

The supermarket would like to be able to predict the likely amount customers in the program will spend, so they can estimate the cost of the rewards. 

This will help them to predict the likely profit at the end of the year.

## Data

The dataset contains records of customers for their last full year of the loyalty program.

| Column Name | Criteria                                                |
|-------------|---------------------------------------------------------|
|customer_id | Unique identifier for the customer. </br>Missing values are not possible due to the database structure. |
|spend | Continuous. </br>The total spend of the customer in their last full year. This can be any positive value to two decimal places. </br>Missing values should be replaced with 0. |
|first_month | Continuous. </br>The amount spent by the customer in their first month of the year. This can be any positive value, rounded to two decimal places. </br>Missing values should be replaced with 0. |
| items_in_first_month | Discrete. </br>The number of items purchased in the first month. Any integer value greater than or equal to zero. </br>Missing values should be replaced by 0. |  
| region | Nominal. </br>The geographic region that the customer is based in. One of four values Americas, Asia/Pacific, Europe, Middle East/Africa. </br>Missing values should be replaced with "Unknown". |
| loyalty_years | Oridinal. </br>The number of years the customer has been a part of the loyalty program. One of five ordered categories, '0-1', '1-3', '3-5', '5-10', '10+'. </br>Missing values should be replaced with '0-1'.|
| joining_month | Nominal. </br>The month the customer joined the loyalty program. One of 12 values "Jan", "Feb", "Mar", "Apr", etc. </br>Missing values should be replaced with "Unknown".|
| promotion | Nominal. </br>Did the customer join the loyalty program as part of a promotion? Either 'Yes' or 'No'. </br>Missing values should be replaced with 'No'.|


# Data Cleaning

Before I fit any models, I need to make sure the data is clean.

The table below shows what the data should look like, so I’ll use that as my guide.

I’ll create a cleaned version of the dataframe.

- I’ll start with the data in the file "loyalty.csv".

- My output will be a dataframe named clean_data.

- I’ll make sure all column names and values match the table below.

| Column Name | Criteria                                                |
|-------------|---------------------------------------------------------|
|customer_id | Unique identifier for the customer. </br>Missing values are not possible due to the database structure. |
|spend | Continuous. </br>The total spend of the customer in their last full year. This can be any positive value to two decimal places. </br>Missing values should be replaced with 0. |
|first_month | Continuous. </br>The amount spent by the customer in their first month of the year. This can be any positive value, rounded to two decimal places. </br>Missing values should be replaced with 0. |
| items_in_first_month | Discrete. </br>The number of items purchased in the first month. Any integer value greater than or equal to zero. </br>Missing values should be replaced by 0. |  
| region | Nominal. </br>The geographic region that the customer is based in. One of four values Americas, Asia/Pacific, Europe, Middle East/Africa. </br>Missing values should be replaced with "Unknown". |
| loyalty_years | Oridinal. </br>The number of years the customer has been a part of the loyalty program. One of five ordered categories, '0-1', '1-3', '3-5', '5-10', '10+'. </br>Missing values should be replaced with '0-1'.|
| joining_month | Nominal. </br>The month the customer joined the loyalty program. One of 12 values "Jan", "Feb", "Mar", "Apr", etc. </br>Missing values should be replaced with "Unknown".|
| promotion | Nominal. </br>Did the customer join the loyalty program as part of a promotion? Either 'Yes' or 'No'. </br>Missing values should be replaced with 'No'.|

In [2]:
# Use this cell to write your code for Task 1
import pandas as pd

clean_data = pd.read_csv('loyalty.csv')

print(clean_data.head())

clean_data['spend'] = clean_data['spend'].round(2)
clean_data['spend'] = clean_data['spend'].fillna(0)
clean_data['first_month'] = clean_data['first_month'].str.replace(r'^\.+$', '0', regex=True)
clean_data['first_month'] = clean_data['first_month'].astype(float)
clean_data['first_month'] = clean_data['first_month'].round(2)
clean_data['items_in_first_month'] = clean_data['items_in_first_month'].fillna(0)
clean_data['region'] = clean_data['region'].fillna('Unknown')
clean_data['joining_month'] = clean_data['joining_month'].fillna('Unknown')
clean_data['loyalty_years'] = clean_data['loyalty_years'].fillna('0-1')
clean_data['loyalty_years'] = pd.Categorical(
    clean_data['loyalty_years'],
    categories=['0-1', '1-3', '3-5', '5-10', '10+'], 
    ordered=True
)
clean_data['promotion'] = clean_data['promotion'].fillna('No')
clean_data['promotion'] = clean_data['promotion'].str.capitalize()
print(clean_data['promotion'].value_counts())

print(clean_data.info())

   customer_id   spend first_month  ...  loyalty_years joining_month promotion
0            1  132.68        15.3  ...           5-10           Nov        No
1            2  106.45        16.2  ...            0-1           Feb       Yes
2            3  123.16       25.26  ...            10+           Dec       Yes
3            4  130.60       24.74  ...            3-5           Apr        No
4            5  130.41       25.59  ...            3-5           Apr       Yes

[5 rows x 8 columns]
No     635
Yes    611
Name: promotion, dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1246 entries, 0 to 1245
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype   
---  ------                --------------  -----   
 0   customer_id           1246 non-null   int64   
 1   spend                 1246 non-null   float64 
 2   first_month           1246 non-null   float64 
 3   items_in_first_month  1246 non-null   int64   
 4   region                1246 no

# Initial Insights 

The team at International Essentials have told me that they have always believed that the number of years in the loyalty scheme is the biggest driver of spend.

I'll produce a table showing the difference in the average spend by number of years in the loyalty programme, along with the variance, to investigate this question for the team.

- I'll start with the data in the file "loyalty.csv".

- My output will be a data frame named spend_by_years.

- It will include the three columns: loyalty_years, avg_spend, and var_spend.

- I'll round all answers to 2 decimal places.   

In [3]:
# Use this cell to write your code for Task 2
import pandas as pd

spend_by_years = pd.read_csv('loyalty.csv')
spend_by_years['loyalty_years'] = spend_by_years['loyalty_years'].fillna('0-1')
spend_by_years = spend_by_years.groupby('loyalty_years')['spend'].agg(['mean', 'var'])
spend_by_years = spend_by_years.reset_index()
spend_by_years.columns = ['loyalty_years','avg_spend', 'var_spend']
spend_by_years = spend_by_years.round(2)
print(spend_by_years)

  loyalty_years  avg_spend  var_spend
0           0-1     110.56       9.30
1           1-3     129.31       9.65
2           10+     117.41      16.72
3           3-5     124.55      11.09
4          5-10     135.15      14.10


# Baseline Model

I have fit a baseline model to predict the spend over the year for each customer.

- I have fit my model using the data contained in "train.csv".

- I have used "test.csv" to predict new values based on my model. I have returned a dataframe named base_result, that includes customer_id and spend. The spend column contains my predicted values.

In [4]:
# Use this cell to write your code for Task 3
import pandas as pd
import seaborn as sns
from sklearn.linear_model import LinearRegression

train = pd.read_csv('train.csv')
X = train.drop('spend', axis=1)
y = train['spend']
test = pd.read_csv('test.csv')
X_test = test
reg = LinearRegression()
reg.fit(X[['first_month', 'items_in_first_month']], y)
predictions = reg.predict(X_test[['first_month', 'items_in_first_month']])
print(predictions.shape)

base_result = pd.DataFrame()

base_result['customer_id'] = test['customer_id']
base_result['spend'] = predictions

print(base_result)

#####################

train_dummies = pd.get_dummies(train, drop_first=True) 
X = train_dummies.drop('sale_price', axis=1)
y = train_dummies['sale_price']
test = pd.read_csv('validation.csv')
test = test.drop('sale_date', axis=1)
X_to_pred = pd.get_dummies(test, drop_first=True)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) 

scaler = StandardScaler() 
X_train_scaled = scaler.fit_transform(X_train) 
X_test_scaled = scaler.transform(X_test)
X_to_pred_scaled = scaler.transform(X_to_pred)

models = {
    'Linear Regression': LinearRegression(),
    'Decision Tree (depth=2)': DecisionTreeRegressor(max_depth=2, random_state=42),
    'Random Forest': RandomForestRegressor(n_estimators=100, random_state=42),
    'Gradient Boosting': GradientBoostingRegressor(n_estimators=100, random_state=42),
}

results = []

for name, model in models.items():
    # Train
    model.fit(X_train_scaled, y_train)
    
    # Predict
    y_pred = model.predict(X_test_scaled)
    
    # Evaluate
    mae = mean_absolute_error(y_test, y_pred)
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    r2 = r2_score(y_test, y_pred)
    
    results.append({
        'Model': name,
        'MAE': mae,
        'RMSE': rmse,
        'R²': r2
    })

print(results)

linreg = LinearRegression()

linreg.fit(X_train_scaled, y_train)
y_pred_final = linreg.predict(X_to_pred_scaled)

compare_result = pd.DataFrame()
compare_result['house_id'] = test['house_id']
compare_result['price'] = y_pred_final

print(compare_result)


(250,)
     customer_id       spend
0              5  129.918426
1              7  130.008029
2             16  129.597136
3             17  131.940284
4             19  132.733573
..           ...         ...
245         1216  132.462654
246         1225  130.327616
247         1231  128.975131
248         1242  129.980938
249         1243  130.386315

[250 rows x 2 columns]


# Compariso Model

I have fit a comparison model to predict the spend over the year for each customer.

- I have fit my model using the data contained in "train.csv".

- I have used "test.csv" to predict new values based on my model. I have returned a dataframe named compare_result, that includes customer_id and spend. The spend column contains my predicted values.

In [13]:
# Use this cell to write your code for Task 4
import pandas as pd
import numpy as np
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.preprocessing import StandardScaler 
from sklearn.model_selection import KFold, train_test_split, GridSearchCV


train = pd.read_csv('train.csv')
train_dummies = pd.get_dummies(train, drop_first=True) 
X = train_dummies.drop('spend', axis=1)
y = train_dummies['spend']
test = pd.read_csv('test.csv')
X_to_pred = pd.get_dummies(test, drop_first=True)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=12) 

scaler = StandardScaler() 
X_train_scaled = scaler.fit_transform(X_train) 
X_test_scaled = scaler.transform(X_test)
X_to_pred_scaled = scaler.transform(X_to_pred)

models = {
    'Linear Regression': LinearRegression(),
    'Decision Tree (depth=2)': DecisionTreeRegressor(max_depth=2, random_state=42),
    'Random Forest': RandomForestRegressor(n_estimators=100, random_state=42),
    'Gradient Boosting': GradientBoostingRegressor(n_estimators=100, random_state=42),
}

results = []

for name, model in models.items():
    # Train
    model.fit(X_train_scaled, y_train)
    
    # Predict
    y_pred = model.predict(X_test_scaled)
    
    # Evaluate
    mae = mean_absolute_error(y_test, y_pred)
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    r2 = r2_score(y_test, y_pred)
    
    results.append({
        'Model': name,
        'MAE': mae,
        'RMSE': rmse,
        'R²': r2
    })

print(results)

linreg = LinearRegression()

linreg.fit(X_train_scaled, y_train)
y_pred_final = linreg.predict(X_to_pred_scaled)

compare_result = pd.DataFrame()
compare_result['customer_id'] = test['customer_id']
compare_result['spend'] = y_pred_final

print(compare_result)


[{'Model': 'Linear Regression', 'MAE': 0.2053887259830745, 'RMSE': 0.28550815852108247, 'R²': 0.9998882071235949}, {'Model': 'Decision Tree (depth=2)', 'MAE': 3.8575350131752315, 'RMSE': 4.698194327712569, 'R²': 0.9697281448142996}, {'Model': 'Random Forest', 'MAE': 0.29876149999999585, 'RMSE': 0.471165824153234, 'R²': 0.9996955439224832}, {'Model': 'Gradient Boosting', 'MAE': 0.35127022531837787, 'RMSE': 0.4922385124779024, 'R²': 0.9996677015893535}]
     customer_id       spend
0              5  140.706932
1              7  148.702688
2             16  140.840268
3             17  150.669489
4             19  153.632073
..           ...         ...
245         1216  134.869847
246         1225  148.530450
247         1231  136.970908
248         1242  129.881431
249         1243  148.515740

[250 rows x 2 columns]


# Conclusion

Throughout this project, I successfully cleaned and prepared the loyalty data, investigated the relationship between loyalty tenure and customer spend, and built both baseline and comparison predictive models. The Linear Regression model delivered the strongest performance with an R² of 0.9999, MAE of 0.2054, and RMSE of 0.2855, significantly outperforming the Decision Tree (R² = 0.9697), Random Forest (R² = 0.9997), and Gradient Boosting (R² = 0.9997). These results confirm the team's long-held belief at International Essentials that the number of years in the loyalty scheme is indeed a powerful driver of spend, and I would recommend the Linear Regression model for predicting future customer spend to support targeted marketing and retention strategies.
